WIP: Rigol oscilloscope refactor

diggit opened the pull request:
This is a Rigol Oscilloscope driver refactor.
Related issues #943

**I need testers for most scope families (see below).**

## Refactor
- each family has it's own identifier and can be easily handled individually
- use STL containers instead of plain C arrays
- use RAII for safe(r) resource management

## Supported devices and new features
- DS1000: Broken  :negative_squared_cross_mark: 
  - times out on some commands and FW crashes, needs more investigation
- MSO/DS1000Z: tested :heavy_check_mark:  (DS1054Z with all options, MSO1104Z with all options)
  - LA: :heavy_check_mark: (MSO1000Z only)
  - implemented memory/samplerate configuration: :heavy_check_mark:
  - improved download speed: :heavy_check_mark:
  - improved trigger poll performance/robustness: :heavy_check_mark:
- MSO5000: fixed trigger polling, added LA support
  - analog, samplerate control, triggering, polling, data acq.: :heavy_check_mark:
  - LA: :heavy_check_mark:
- DHO800:
  - analog, samplerate control, triggering, polling, data acq.: :heavy_check_mark:
  - switched download to single block, doubled the download speed
- DHO900: no intended changes, but untested (I can borrow one to validate my changes) :pray:
  - should be the same as DHO800
  - LA: untested
- DHO1000: no intended changes, but untested :pray: 
- DHO4000: no intended changes, but untested :pray:

## Known issues (TBD):
- Memory depth does not update (in ngscopesclient)  when channels are added to the capture. https://github.com/ngscopeclient/scopehal-apps/issues/920
- MSO5000 trigger detection is unreliable  (on 8 gs/s we may completely miss the non "STOP" state)
- Driver uses it's own sample conversion function, instead of [one from core](https://github.com/ngscopeclient/scopehal/issues/1035#issuecomment-3700530441)  
- use `SetADCMode`/`GetADCMode` for devices which support multiple sample resolutions (at least DHOs - those current use custom settings property)

## non device specific
- Some trigger offset fixes (keep in place and in valid range on samplerate/memory depthy changes)

### Triggers
- Pulse width: WIP (#1019)

## Not _yet_ supported devices
- MSO/DS2000: not supported yet
- MSO/DS4000: not supported yet (I have access to this device and can work on the support)
- MHO900/MHO098: not supported yet, device available via network (#1022)

diggit committed:
RigolOscilloscope: major driver refactor

Differentiate between scope families and
then group their common behavior in switch statements.
This is more flexible approach and allow more fine tuned,
per family behavior than immediately grouping them by protocol.

- use smart pointers where it makes sense
- use vector/array rather than C arrays

DS1000Z
- implemented memory/samplerate configuration
- improved download speed
- improved trigger poll performance/robustness

Other families should have same handling as before.

diggit committed:
RigolOscilloscope: enforce all case handling in switch statements

to make sure all scope families are handled everywhere

diggit committed:
RigolOscilloscope: clamp trigger offset into valid range

When trigger offset was beyond memory range,
at least DS1000Z never received trigger.

diggit committed:
RigolOscilloscope: mark trigger offset as invalid when new values is set

diggit committed:
RigolOscilloscope: update trigger offset on samplerate change

This is necessary, because samplerate changes timebase.
While TB changes, Rigol oscilloscopes keep the trigger position constant
relative to the capture mid point.
Because scopehal reference point is at start, the trigger has to be updated
to keep it constant from scopehal's POV.

diggit committed:
RigolOscilloscope: mark interleaving as not supported

As interleaving done automatically by scope is not considered
to be an interleaving feature, see #1014

diggit committed:
RigolOscilloscope: fix DS1000Z max download chunk size (maxpoints)

Initial tests with 1 channel and 1 MB chunks worked fine.
Later tests with 2 channels failed, but 500 kB chunk size worked fine,
but failed for 3 or 4 channels. It looks like the chunk size (maxpoints),
also depends on the amount of enabled channels.
There may be still some margin, but it's not worth chaseing it.

diggit committed:
RigolOscilloscope: refactor resource locking

Proper resource locking was missing at multiple places.
It may not be 100% complete yet.

diggit committed:
RigolOscilloscope: clean-up and extend logging

- change verbose logs to trace
- add more trace logs
- print detected options
- remove purely development logs

diggit committed:
RigolOscilloscope: fix `GetChannelBandwidthLimiters` for DHO series

There are DHO4000 series scopes, which support 250 MHZ BW limit.

diggit committed:
RigolOscilloscope: simplify channel bw limit handling

- all the series/model -> limiters logic is now in `GetChannelBandwidthLimiters`
- `SetChannelBandwidthLimit` uses values from `GetChannelBandwidthLimiters` and uses nearest higher limit

diggit committed:
RigolOscilloscope: fix DS1000Z samplerate/timebase conversion rules

it is more complicated than expected, no obvious rules

diggit committed:
RigolOscilloscope: DS1000Z: allow 24M point memory also for multiple channels

diggit committed:
RigolOscilloscope: deduplicate pre/post start code

diggit committed:
RigolOscilloscope: DS1000Z: improve trigger handling

- deduce whole trigger state from data preamble rather than status
- resilient against fast triggers skipping transitions
- properly handles busy(wait)/armed(run) states

diggit committed:
RigolOscilloscope: minor cleanups

diggit committed:
RigolOscilloscope: change `Series::DS1000Z` enumeration to change Series::MSODS1000Z as there are also MSO scopes

diggit committed:
RigolOscilloscope: enable channel VERNier for all series

MSODS1000Z was the only exception and it works there.

diggit committed:
RigolOscilloscope: fix serires detection

- recognize MSO1000Z as MSODS1000Z (tested)
- recognize DHO800/DHO900/DHO1000/DHO4000 (untested)

diggit committed:
RigolOscilloscope: add response parsing hekpers

only `double` for now, but templating allows easy addition of more

diggit committed:
RigolOscilloscope: add RAII "End Of Context" helper to make resource cleanups easier

diggit committed:
RigolOscilloscope: move initial `UpdateDynamicCapabilities` after channel configuration

- it may touch channels and it caused bugs when adding LA support

diggit committed:
RigolOscilloscope: reduce includes

most of necessary includes are already included indirectly

diggit committed:
RigolOscilloscope: add support for LA

- developed with MSO1000Z series
- MSO5000/DHO900 implemented, but not tested

diggit committed:
RigolOscilloscope: change all cached bools (`value` + `ValueValid`) to `std::optional`

diggit committed:
RigolOscilloscope: implement `CanEnableChannel` and use it in EnableChannel

- ATM only affects MSODS1000Z as there are collisions between digital banks and some analog channels

diggit committed:
RigolOscilloscope: rename `GetChannelDivisor` -> `GetMsods1000ZChannelDivisor`

because ATM this divisor behavior seems to be specific to MSODS1000Z only

diggit committed:
RigolOscilloscope: MSODS1000Z: fix possible bad locking in `SetSampleDepth`

- we have to ensure, the trigger status does not change after we query it
so we can restore correct value in the end

diggit committed:
RigolOscilloscope: create external trigger on for series with EXT trig input

- all supported except MSO5000

diggit committed:
RigolOscilloscope: add AC line (ACL) trigger,

scopehal does not support it now,
so it will not be listed in available trigger inputs

diggit committed:
RigolOscilloscope: simplify `EnableChannel` immediate effect quirk, add MSO5000 as affected

diggit committed:
RigolOscilloscope: discard cached srate value when`EnableChannel`/`DisableChannel` is called

number of enabled channels may affect current samplerate, especially near the maximum rates

diggit committed:
RigolOscilloscope: MSO5000: fix samplerate setting

- it is dependent on how many analog channels are enabled

diggit committed:
RigolOscilloscope: logging , comment/TODO and helper function usage cleanup

diggit committed:
RigolOscilloscope: `PollTrigger`: add quirk for MSO5000

to not report `TRIGGER_MODE_TRIGGERED` when scope state == `TD`,
because then the `ngscopeclient` immediately calls `AcquireData`,
but data are available once the scope reaches `STOP` state, not `TD`.

diggit committed:
RigolOscilloscope: make `AcquireData` more resistant to missing samplerate info in preamble

diggit committed:
RigolOscilloscope: `SetSampleRate`: ignore unsupported samplerate requests

diggit committed:
RigolOscilloscope: add missing `GetMso5000AnalogBankUsage` declaration

azonenberg commented:
@diggit where are we on this, is it ready to review and merge? Are we waiting on more testing? I'd like to get this in for v0.1.1 if possible.

azonenberg commented:
If you want to target 0.2 that's fine. I don't want to push you just checking in on status.

diggit committed:
RigolOscilloscope: minor spelling/logging fixes

diggit committed:
RigolOscilloscope: only 2 ch. versions of DHO800 have EXT trigger

diggit committed:
RigolOscilloscope: DHO800 have max 25M memory depth

note from release notes was misinterpretted 5M != 50M

diggit committed:
RigolOscilloscope: `SetSampleDepth(...)` : ignore sample depths not in `GetSampleDepthsNonInterleaved()`

(Rigol scopes do not support interleaved mode)

diggit committed:
RigolOscilloscope: `SetSampleRate(...)`: clear `m_mdepth` only for relevant series

and is is not clear, it is is even strictly necessary

diggit committed:
RigolOscilloscope: `SetSampleDepth(...)`: simplify for DHO series

no need to match for supported rates as this is already checked at the beginning

diggit committed:
RigolOscilloscope: `SetSampleRate(...)`: add DHO quirk to immediately take effect

At least on DHO800 series, the scope has to do at least one trigger after new samplerate (time base) is set,
before it starts returning thenew value.
RUN mode is not enough in this case.

diggit committed:
RigolOscilloscope: `SetTriggerOffset(...)`:  DHO series: push new trigger time offset in RUN mode

When in STOP, device limits the trigger offset changes to very limited range (at least DHO800)

diggit committed:
RigolOscilloscope: `GetExternalTrigger()`:  return the m_extTrigChannel

not sure why it always returned `nullptr` unconditionally

diggit committed:
RigolOscilloscope:  MSO5000 and DHO series: download waveform data all at once

these scopes can send all at once making the transfer significantly faster (half the time)

diggit committed:
RigolOscilloscope:  unify data download for MSO5000/MSODS1000/DHO series

the only difference are legacy DS1000 series

diggit committed:
RigolOscilloscope: fix MSODS1000Z data download

in recent commit, `maxpoints` got always updated from preamble, while it should happen only when it is zero

diggit committed:
RigolOscilloscope: update parsing heleprs to use `std::optional<T>`

this way, we can have invalid value even for integral types

diggit committed:
RigolOscilloscope: when `GetSampleDepth()` fails to parse value, try to set lowest known mdepth and re-read

this may happen if the scope memory depth i in AUTO mode and we can't work with that

diggit committed:
RigolOscilloscope: default initialize all primitive member data

for safety reasons

diggit committed:
RigolOscilloscope: some DS1000E notes, remains broken ATM

diggit committed:
RigolOscilloscope: `GetSampleDepth` do not crash when mdepths are unknown

diggit committed:
RigolOscilloscope: fix: always parse sample memory depth as double

diggit committed:
RigolOscilloscope: disable ato roll mode for DHOs

diggit committed:
RigolOscilloscope: `PollTrigger()` enable `TD` state masking for all series

because at least `ngscopeclient` starts the download when the scope enters triggered state, but the data are not ready yet.
So we interpret `TD` as `TRIGGER_MODE_RUN` and flip to triggered/stop later

diggit committed:
RigolOscilloscope: default initialize all member data of `CapturePreamble`

just in case so we don't read some garbage

diggit commented:
DHO800 was mostly working. Some quirks were added.
DS1000E support is broken. Initial handshake fails as the cope does not respond to `:CHAN<N>:DISP?` commands (it should).

Wiljan1 commented:
MSO5072 with upgrade for 4 channels and 200MHz here.
Only CH1+2 does show up in ngscopeclient-windows-x64-v0.1.msi  on win10
Also get an error message when I hover over the output of ch2
![Rigol_ngscopeclient_error](https://github.com/user-attachments/assets/02d6b10e-4b75-4038-98cd-ddbe49cd0c54)
Look like other see same behavior [https://www.eevblog.com/forum/testgear/ngscopeclient-ultimate-acquisition-software-protocol-decoder-debug-tool/50/](url)

Wiljan1 commented:
Hi @diggit ,  

`I am not aware of any way how to inject some per instance overrides, so I will look into some alternative way, how to detect available channels.
`

I think the :SYSTem:OPTion:STATus? can be used to check number of Analog channels, BW :

`:SYSTem:OPTion:STATus?
Syntax :SYSTem:OPTion:STATus? <type>
Description Queries the status of the options of the oscilloscope.
Parameter Name Type Range Default
<type> Discrete
{BW071|BW072|BW073|BW12|BW13|BW
23|RL2|4CH|BND|COMP|EMBD|AUTO|FLE
X|AUDIO|AERO|AWG|PWR}
——
说明 ➢ BW071: bandwidth upgrade option. The bandwidth upgrades from 70 MHz to 100 MHz.
➢ BW072: bandwidth upgrade option. The bandwidth upgrades from 70 MHz to 200 MHz.
➢ BW073: bandwidth upgrade option. The bandwidth upgrades from 70 MHz to 350 MHz.
➢ BW12: bandwidth upgrade option. The bandwidth upgrades from 100 MHz to 200 MHz.
➢ BW13: bandwidth upgrade option. The bandwidth upgrades from 100 MHz to 350 MHz.
➢ BW23: bandwidth upgrade option. The bandwidth upgrades from 200 MHz to 350 MHz.
➢ RL2: memory depth option. Maximum memory depth up to 200 Mpts.
➢ 4CH: channel number upgrade option. Upgrade the number of analog channels to 4
Chapter 2 Command System RIGOL
MSO5000 Programming Guide 2-169
(only available for the MSO5XX2 model).
➢ BND: function and application bundle option, MSO5000-COMP, MSO5000-EMBD,
MSO5000-AUTO, MSO58000-FLEX, MSO5000-AUDIO, MSO5000-AERO, MSO5000-AWG,
and MSO5000-PWR.
➢ COMP: PC serial bus trigger and analysis option (RS232/UART).
➢ EMBD: embedded serial bus trigger and analysis option (I2C and SPI).
➢ AUTO: auto serial bus trigger and analysis option (CAN and LIN).
➢ FLEX: FlexRay serial bus trigger and analysis option (FlexRay).
➢ AUDIO: audio serial bus trigger and analysis (I2S, only available for the MSO5XX4 model
or the model installed with the MSO5000-4CH option).
➢ AERO: MIL-STD-1553 serial bus trigger and analysis option (MIL-STD-1553).
➢ AWG: dual-channel 25 MHz Arbitrary Waveform Generator.
➢ PWR: built-in power analysis option.
Return
Format
The query returns 0 or 1.
0: indicates that the option is not installed.
1: indicates that the official option has been installed.`

Let me know if you want me to test something?

Thank you
Wiljan1

diggit commented:
@Wiljan1 I've tried `:SYSTem:OPTion:STATus?`/`:SYST:OPT:STAT?` on my MSO5074 and it times out without response.
Alternative approach could be to request `:CHAN3:DISP 1` and then check for an error. Tried that with `CHAN5` on mentioned (4ch) scope and it looks feasible, but feedback from someone with MSO5072 without the 4 CH option installed, would be appreciated.

diggit committed:
RigolOscilloscope: add experimental support for MSO5xx2 4CH indirect option detection

MSO5xx2 scopes are scopes with only 2 channels and other 2 channels are SW option.
Usually `:SYST:OPT:STAT? 4CH` query should be enough, but there are modifier FWs
which report `0` (not present), but the option is active.
This is and attempt to detect it using `:CHAN3:VERN ON`,
which is expected to trigger error if option is missing.
We intentionally do not use any CHAN3 query, as those time out.

diggit commented:
@Wiljan1 new commit 20864ba should address 4CH option detection. Please give it a shot on some MSO5xx2 device (preferably with and without this option).
This check is only done for MSO5xx2 scopes. We assume 4 channels on MSO5xx4 models.

Wiljan1 commented:
@diggit thx I need to know how add the commit to the build on windows in the MSYS2 so I can test it, right now I can compile the main git source.

Wiljan1 commented:
@diggit when I add your instructions above and do a build it unfortunate fails:

`[201/542] Building CXX object lib/scopehal/CMakeFiles/scopehal.dir/RigolOscilloscope.cpp.obj
FAILED: [code=1] lib/scopehal/CMakeFiles/scopehal.dir/RigolOscilloscope.cpp.obj

In file included from C:/msys64/home/Jan Jakobsen/scopehal-apps/lib/scopehal/RigolOscilloscope.cpp:31:
C:/msys64/home/Jan Jakobsen/scopehal-apps/lib/scopehal/RigolOscilloscope.h:128:17: warning: declaration of 'BYTE' shadows a global declaration [-Wshadow]
  128 |                 BYTE = 0, // a waveform point occupies one byte (namely 8 bits).
      |                 ^~~~
In file included from C:/msys64/ucrt64/include/windef.h:9,
                 from C:/msys64/ucrt64/include/windows.h:69,
                 from C:/msys64/ucrt64/include/winsock2.h:23,
                 from C:/msys64/ucrt64/include/ws2tcpip.h:17,
                 from C:/msys64/home/Jan Jakobsen/scopehal-apps/lib/scopehal/scopehal.h:42,
                 from C:/msys64/home/Jan Jakobsen/scopehal-apps/lib/scopehal/RigolOscilloscope.cpp:30:
C:/msys64/ucrt64/include/minwindef.h:139:25: note: shadowed declaration is here
  139 |   typedef unsigned char BYTE;
      |                         ^~~~

and a lot more warnings and finally


cc1plus.exe: all warnings being treated as errors
[204/542] Building CXX object lib/scopehal/CM...es/scopehal.dir/TektronixOscilloscope.cpp.ob
ninja: build stopped: subcommand failed.`

diggit committed:
RigolOscilloscope: fix `BYTE` enumeration collision with msys64 typedef

diggit committed:
RigolOscilloscope: resolve variable name shadow warning

diggit committed:
RigolOscilloscope: simplify MSO5xx2 4CH detection using `:SYST:RAM?`

diggit commented:
The [code is there and it's pretty straightforward](https://github.com/ngscopeclient/scopehal/pull/1017/files#diff-c58751e5d8cdd472b168b95b6ffcec1fb33fa423d7720792b8b993769ff8814dR432-R443) (you have to click to display the  `RigolOscilloscope.cpp` diff)

1. check your current commit hash `cd lib; git rev-parse HEAD` is current latest one `bb36776cd48e758b436fe5098a739b25530c1f17`
2. fetch and rebuild if needed
3. use `path/to/ngscopeclient --debug --trace SCPISocketTransport --trace SCPITransport --trace RigolOscilloscope` to get much more info about what is happening (I assume you are using raw TCP socket - port 5555)
4. save the output into file and upload it here (you may want to redact your device serial number)

This is my output, I've overridden channels detected from model to be 2 to try it on MS5074.
```
Connecting to SCPI device at 10.0.0.219:5555
[SCPISocketTransport::SendCommand] [10.0.0.219] Sending *IDN?
[SCPISocketTransport::ReadReply] [10.0.0.219] Got RIGOL TECHNOLOGIES,MSO5074,REDACTEDREDAC,00.01.03.03.00
[RigolOscilloscope::DecodeDeviceSeries] Decoding device series
[RigolOscilloscope::AnalyzeDeviceCapabilities] Analyzing scope capabilities
[SCPISocketTransport::SendCommand] [10.0.0.219] Sending SYST:RAM?
[SCPISocketTransport::ReadReply] [10.0.0.219] Got 4
```
And I get 4 channels in the ngscopeclient in the end.

EDIT: some clarifications

diggit committed:
RigolOscilloscope: fix `WORD` enumeration collision with ucrt64 typedef

diggit committed:
RigolOscilloscope: fix wrong printf types

- so these work on different platforms

dj-bauer commented:
I would also be interested in testing it on my DHO 804 with DHO925S firmware.

I also wonder generally, why most drivers are in 1.5k-3k LOC files instead of spreading them out into different files.
imo this would help spread out the complexity when merging this rework. (I know here are also other things at play, but generally speaking)
If there is a reason for it forget my suggestion.

I would be happy to support with my time to move this PR forward, but this +2700 -800 changelog makes it appear difficult to get a starting point to move from

diggit commented:
> I would also be interested in testing it on my DHO 804 with DHO925S firmware.

You can, just build from my branch.

> I also wonder generally, why most drivers are in 1.5k-3k LOC files instead of spreading them out into different files. imo this would help spread out the complexity when merging this rework. (I know here are also other things at play, but generally speaking) If there is a reason for it forget my suggestion.

I kept all in one source file as it already was. I also don't see a benefit in spliting driver into multiple files.

> I would be happy to support with my time to move this PR forward, but this +2700 -800 changelog makes it appear difficult to get a starting point to move from

Yes, diff is massive because I've changed the style of the driver internals to easily support various series (series specific enum, switches). There is also a lot of new, series specific quirks.

Resolving conflicts will be a bit annoying. I will look into this by the end of the summer, unless someone else decides to take over.

dj-bauer commented:
> You can, just build from my branch.

Do you have a commit in scopehal-apps that works with your branch?
It doesn't seem to build against upstream

diggit commented:
I have https://github.com/ngscopeclient/scopehal-apps/commit/36aca60bc96a4cb9f43b39ab41fe1061ef18569e checked out. That should build.

dj-bauer commented:
Hardware: DHO 804
Quick things I noticed **(DHO 925 firmware)** allthough I don't exactly know wether it's a bug with the DHO firmware or in the drivers yet:

# Timebase samplerate dropdown always one selection behind
- "Arm the trigger in normal mode" -> constantly streaming new samples measuring my scopes callibration output.
- I set the samplerate to 1.25Gs in ngscopeclient. the waveform is properly sampled at 1.25Gs (seeing it on my scopes display and when hovering over the waveform in the viewer. 
- But the dropdown field in the timebase section reset back to 625MS/s This might be related to your open issue about timebase not updating.
- But then I noticed a pattern. While I kept pulling waveforms I selected random waveforms and the input select would always jump back to the previous selection.
- So to my navie brain it appears it actually updates  itself, but it updates itself, before the internal representation of the scope has gotten the new sample rate assigned. Of course there could also be a whole other ratsnest behind it, but this was my first finding using it

I wanted to note this as the issue about timebase mainly focused on channels changing, but this seems to be a ui interaction bug
